### PR TITLE
Improve the behavior of scrollbar drag-scrolls triggered by the trackpad

### LINKED ISSUE #149999: Odd scroll behavior and occasional asserts with trackpad
### Steps to reproduce

1. Run flutter create and replace main.dart with the provided code sample.
2. Run on macOS
3. Do a scroll gesture over the ListView using the trackpad

### Expected results

Smooth scroll with bouncing overscroll effect
Can reverse scroll direction without releasing double tap

### Actual results

Scroll appears to jump erratically, no overscroll effect
When at viewport top, after trying to scroll up further cannot reverse scroll down by reversing the swipe direction 

Current master:
master

After reverting to 529a4d2:
revert

Occasionally the following asserts are seen:

<details open><summary>Logs</summary>

```
════════ Exception caught by scheduler library ═════════════════════════════════
The following assertion was thrown during a scheduler callback:
'package:flutter/src/animation/animation_controller.dart': Failed assertion: line 898 pos 12: 'elapsedInSeconds >= 0.0': is not true.

Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.yml

When the exception was thrown, this was the stack:
#2      AnimationController._tick (package:flutter/src/animation/animation_controller.dart:898:12)
#3      Ticker._tick (package:flutter/src/scheduler/ticker.dart:258:12)
#4      SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1392:15)
#5      SchedulerBinding.handleBeginFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:1235:11)
#6      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
#7      SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:1233:17)
#8      SchedulerBinding._handleBeginFrame (package:flutter/src/scheduler/binding.dart:1150:5)
#9      _invoke1 (dart:ui/hooks.dart:328:13)
#10     PlatformDispatcher._beginFrame (dart:ui/platform_dispatcher.dart:397:5)
#11     _beginFrame (dart:ui/hooks.dart:272:31)
(elided 2 frames from class _AssertionError)

This exception was thrown in the context of a scheduler callback. When the scheduler callback was _registered_ (as opposed to when the exception was thrown), this was the stack:
#2      SchedulerBinding.scheduleFrameCallback (package:flutter/src/scheduler/binding.dart:606:49)
#3      Ticker.scheduleTick (package:flutter/src/scheduler/ticker.dart:274:46)
#4      Ticker.start (package:flutter/src/scheduler/ticker.dart:180:7)
#5      AnimationController._startSimulation (package:flutter/src/animation/animation_controller.dart:818:42)
#6      AnimationController._animateToInternal (package:flutter/src/animation/animation_controller.dart:689:12)
#7      AnimationController.forward (package:flutter/src/animation/animation_controller.dart:498:12)
#8      new InkHighlight (package:flutter/src/material/ink_highlight.dart:60:9)
#9      _InkResponseState.updateHighlight (package:flutter/src/material/ink_well.dart:992:29)
#10     _InkResponseState.handleHoverChange (package:flutter/src/material/ink_well.dart:1269:5)
#11     _InkResponseState.handleMouseEnter (package:flutter/src/material/ink_well.dart:1257:7)
#12     MouseTracker._handleDeviceUpdateMouseEvents (package:flutter/src/rendering/mouse_tracker.dart:423:29)
#13     MouseTracker._handleDeviceUpdate (package:flutter/src/rendering/mouse_tracker.dart:277:5)
#14     MouseTracker.updateAllDevices.<anonymous closure> (package:flutter/src/rendering/mouse_tracker.dart:373:9)
#15     MouseTracker._deviceUpdatePhase (package:flutter/src/rendering/mouse_tracker.dart:205:9)
#16     MouseTracker.updateAllDevices (package:flutter/src/rendering/mouse_tracker.dart:367:5)
#17     RendererBinding._scheduleMouseTrackerUpdate.<anonymous closure> (package:flutter/src/rendering/binding.dart:485:22)
#18     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1392:15)
#19     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1326:11)
#20     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1171:5)
#21     _invoke (dart:ui/hooks.dart:312:13)
#22     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:419:5)
#23     _drawFrame (dart:ui/hooks.dart:283:31)
(elided 2 frames from class _FrameCallbackEntry)
════════════════════════════════════════════════════════════════════════════════
```

</details>

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final ScrollController _scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Scroll me'),
            Expanded(
              child: Scrollbar(
                controller: _scrollController,
                child: ScrollConfiguration(
                  behavior: ScrollConfiguration.of(context).copyWith(
                    scrollbars: false,
                    physics: const BouncingScrollPhysics(),
                  ),
                  child: ListView(
                    controller: _scrollController,
                    children: List<ListTile>.generate(50, (index) => ListTile(title: Text('Tile $index'), onTap: () => print('tap! $index'))),
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### LINKED ISSUE #150236: Horizontal `ListView` in nested `ListView`s doesn't work
### Steps to reproduce

1. Run the sample code
2. Scroll the horizontal list view

### Expected results

Should be able to scroll horizontally.

### Actual results

Doesn't scroll and the scroll bar for the vertical list view is highlighted. This issue happens on both iOS and web but only happens when I'm using trackpad.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [Carousel].

void main() => runApp(const ExampleApp());

class ExampleApp extends StatelessWidget {
  const ExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Carousel Sample'),
        ),
        body: const Example(),
      ),
    );
  }
}

class Example extends StatefulWidget {
  const Example({super.key});

  @override
  State<Example> createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  @override
  Widget build(BuildContext context) {
    return ListView(
      children: <Widget>[
        Container(
          width: 200,
          height: 200,
          color: Colors.amber[50],
        ),
        Center(
          child: ConstrainedBox(
            constraints: const BoxConstraints(maxHeight: 200),
            child: ListView(
              scrollDirection: Axis.horizontal,
              children: List.generate(20, (int index) {
                return ConstrainedBox(constraints: BoxConstraints.tightFor(width: 200), child: Card.outlined(child: SizedBox(width: 200, height: 200, child: Center(child: Text('Item $index')),),),);
              }),
            )
          ),
        ),
        Container(
          width: 200,
          height: 200,
          color: Colors.amber[100],
        ),
        Container(
          width: 200,
          height: 200,
          color: Colors.amber[200],
        ),
      ],
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/36861262/fbb2f7b8-adf2-4dbe-8388-8c35464d4b1e

</details>

Might be related to https://github.com/flutter/flutter/pull/146654#

CC: @HansMuller

### LINKED ISSUE #150342: Scrollbar gesture recognizers not updated on scroll extent change
### Steps to reproduce

1. Run flutter create and replace main.dart with the provided code sample.
2. Run on macOS
3. Wait a few seconds for the text in the TextField to appear
4. Do a scroll gesture over the Text using the trackpad
5. Move the mouse pointer outside of the app window or tap on the app bar
6. Again, do a scroll gesture over the Text using the trackpad

### Expected results

Consistent trackpad scroll behaviour, identical list of gesture recognizers.

### Actual results

Inconsistent trackpad scroll behaviour, different list of gesture recognizers.

Before text update:
```
flutter: Gesture arena 2    ❙ ★ Opening new gesture arena.
flutter: Gesture arena 2    ❙ Adding: VerticalDragGestureRecognizer#8b17a(start behavior: start)
flutter: Gesture arena 2    ❙ Closing with 1 member.
flutter: Gesture arena 2    ❙ Default winner: VerticalDragGestureRecognizer#8b17a(start behavior: start)
```

After text update:
```
flutter: Gesture arena 4    ❙ ★ Opening new gesture arena.
flutter: Gesture arena 4    ❙ Adding: VerticalDragGestureRecognizer#8b17a(start behavior: start)
flutter: Gesture arena 4    ❙ Adding: _VerticalThumbDragGestureRecognizer#293c1(debugOwner: _MaterialScrollbarState#10516(tickers: tracking 2 tickers), start behavior: down)
flutter: Gesture arena 4    ❙ Adding: _VerticalThumbDragGestureRecognizer#eacad(debugOwner: _MaterialScrollbarState#1f937(tickers: tracking 2 tickers), start behavior: down)
flutter: Gesture arena 4    ❙ Closing with 3 members.
flutter: Gesture arena 4    ❙ Accepting: _VerticalThumbDragGestureRecognizer#293c1(debugOwner: _MaterialScrollbarState#10516(tickers: tracking 2 tickers), start behavior: down)
flutter: Gesture arena 4    ❙ Self-declared winner: _VerticalThumbDragGestureRecognizer#293c1(debugOwner: _MaterialScrollbarState#10516(tickers: tracking 2 tickers), start behavior: down)
```

scroll-issue-async

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/gestures.dart' show debugPrintGestureArenaDiagnostics;
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final ScrollController _scrollController = ScrollController();
  final TextEditingController _textEditingController = TextEditingController();

  @override
  void initState() {
    super.initState();
    debugPrintGestureArenaDiagnostics = true;

    Future.delayed(const Duration(milliseconds: 3000), () {
      _textEditingController.text =
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt '
          'ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco '
          'laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in '
          'voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat '
          'non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\n\n' * 10;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          const Text('Wait for it...'),
          Expanded(
            child: Scrollbar(
              controller: _scrollController,
              child: ScrollConfiguration(
                behavior: ScrollConfiguration.of(context).copyWith(
                  scrollbars: false,
                  physics: const BouncingScrollPhysics(),
                ),
                child: TextField(
                  controller: _textEditingController,
                  scrollController: _scrollController,
                  maxLines: null,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
% flutter doctor -v
[✓] Flutter (Channel master, 3.23.0-13.0.pre.242, on macOS 14.5 23F79 darwin-arm64, locale en-ES)
    • Flutter version 3.23.0-13.0.pre.242 on channel master at /Users/tgucio/Library/Frameworks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5187cab7bd (2 days ago), 2024-06-14 23:35:34 -0400
    • Engine revision 9779c273aa
    • Dart version 3.5.0 (build 3.5.0-266.0.dev)
    • DevTools version 2.36.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/tgucio/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.90.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.5 23F79 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.5 23F79 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by HansMuller**
Corrects some problems related to trackpad scrolls introduced by https://github.com/flutter/flutter/pull/146654.

Fixes https://github.com/flutter/flutter/issues/149999
Fixes #150342
Fixes https://github.com/flutter/flutter/issues/150236

### CODE REVIEW

**Comment by mark8044** [comment]
Is this going into the master branch? Would love to test this as well

**Comment by HansMuller** [comment]
Yes, the plan is to land this PR on the master branch.

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -1324,6 +1324,7 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
   final GlobalKey  _scrollbarPainterKey = GlobalKey();
   bool _hoverIsActive = false;
   Drag? _thumbDrag;
+  bool _isScrollable = false;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1327]: It does not look like this is used for anything, can it be removed?

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -1869,6 +1867,9 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
     if (metrics.axis != _axis) {
       setState(() { _axis = metrics.axis; });
     }
+    if (_isScrollable != notification.metrics.maxScrollExtent > 0) {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1870]: This also is not necessarily true. Some scroll physics allow scrolling when there is no scrollable extent.

File: packages/flutter/test/widgets/scrollbar_test.dart [hunk 3]
```diff
@@ -3240,20 +3240,191 @@ The provided ScrollController cannot be shared by multiple ScrollView widgets.''
     expect(scrollController.offset, 0.0);
```
Comment (Piinks) [inline packages/flutter/test/widgets/scrollbar_test.dart:3240]: Is there a regression test for https://github.com/flutter/flutter/issues/150342?

File: packages/flutter/lib/src/widgets/scrollbar.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/scrollbar.dart:1870]: I made the name too short. It should really be called maxScrollExtentPermitsScrolling; it's only used to detect this one state change and trigger a rebuild.

File: packages/flutter/test/widgets/scrollbar_test.dart (hunk 3 above)
Comment (HansMuller) [reply packages/flutter/test/widgets/scrollbar_test.dart:3240]: Good point, there isn't one. I will add it.

File: packages/flutter/lib/src/widgets/scrollbar.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/scrollbar.dart]: done

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 4]
```diff
@@ -1915,16 +1915,26 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
     handleThumbPress();
   }
 
+  // The protected RawScrollbar API methods - handleThumbPressStart,
+  // handleThumbPressUpdate, handleThumbPressEnd - all depend on a
+  // localPosition parameter that defines the event's location relative
+  // to the scrollbar.  Ensure that the localPosition is reported consistently,
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/scrollbar.dart]: Thanks, done.

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 5]
```diff
@@ -1966,7 +1976,6 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
     if (!_canHandleScrollGestures()) {
       return gestures;
     }
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1978]:
```suggestion
    }

```

**Review by Piinks — APPROVED** [review]
LGTM

**Pull request merged**